### PR TITLE
feat(cluster-overview): add nodes table

### PR DESCRIPTION
**Pull request by RemiBonnet**
# What does this PR do?

https://qovery.atlassian.net/jira/software/c/projects/QOV/boards/56?selectedIssue=QOV-814

- Update `ProgressBar` adding `mode` behavior
- Add new request `useClusterMetrics` instead using the data from `useClusterRunningStatus` 
- Add nodepool and node tables (only node tables for non-Karpenter clusters)
- Add [warnings](https://qovery.slack.com/archives/C0811LV69C7/p1748866371940749) elements

Cluster - Overview · 4 42pm · 06-04

---

## PR Checklist

- [ ] This PR introduces breaking change(s) and has been labeled as such
- [ ] This PR introduces new store changes
- [x] I made sure the code is type safe (no any)

### CODE REVIEW

**Comment by RemiBonnet** [comment]
[Qovery Preview](https://www.qovery.com)
---

Qovery can create a Preview Environment for this PR.
To trigger its creation, please post a comment with one of the following command.

| Command | Blueprint environment |
|---|---|
| /qovery []() preview cc1de7de-94e5-40f8-8e45-bc8986d9dfec | [storybook](https://console.qovery.com/organization/3d542888-3d2c-474a-b1ad-712556db66da/project/a64f6cc1-4e34-433b-b8fd-5359716fb01b/environment/cc1de7de-94e5-40f8-8e45-bc8986d9dfec/services/general) |
 /qovery []() preview 28c47145-c8e7-4b9d-8d9e-c65c95b48425 | [staging](https://console.qovery.com/organization/3d542888-3d2c-474a-b1ad-712556db66da/project/a64f6cc1-4e34-433b-b8fd-5359716fb01b/environment/28c47145-c8e7-4b9d-8d9e-c65c95b48425/services/general) |
| /qovery []() preview {all\|UUID1,UUID2,...} | To preview multiple environments |

---

This comment has been generated from [Qovery](https://qovery.com) AI :robot:.
Below, a word from its wisdom :
> One day they will all understand

File: libs/domains/cluster-metrics/feature/src/lib/cluster-table-nodepool/calculate-nodepool-metrics.spec.ts [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+    cpu_milli_limit: 0,
+    memory_mib: 0,
+    memory_mib_limit: 0,
+    nodes_count: 0,
+  }
+
+  const mockStableNodePool: NodePoolInfoDto = {
+    name: 'stable',
+    cpu_milli: 2000,
+    cpu_milli_limit: 4000,
```
Comment (rmnbrd) [inline libs/domains/cluster-metrics/feature/src/lib/cluster-table-nodepool/calculate-nodepool-metrics.spec.ts:180]: @RemiBonnet Could you add another unit test to test the case where `cpu_milli_limit` is `null` ? 🙏

File: libs/domains/clusters/feature/src/lib/cluster-card/cluster-card.tsx [hunk 2]
```diff
@@ -82,9 +89,15 @@ export interface ClusterCardProps {
 export function ClusterCard({ cluster, clusterDeploymentStatus }: ClusterCardProps) {
   useClusterRunningStatusSocket({ organizationId: cluster.organization.id, clusterId: cluster.id })
 
+  const isRunningStatusEnabled = useFeatureFlagVariantKey('cluster-running-status')
```
Comment (rmnbrd) [inline libs/domains/clusters/feature/src/lib/cluster-card/cluster-card.tsx]:
```suggestion
  const isClusterOverviewEnabled = useFeatureFlagVariantKey('cluster-running-status')
```

File: libs/pages/cluster/src/lib/ui/container/container.tsx [hunk 3]
```diff
@@ -11,19 +13,21 @@ import {
...
 import { IconEnum } from '@qovery/shared/enums'
-import { CLUSTER_SETTINGS_URL, CLUSTER_URL } from '@qovery/shared/routes'
+import { CLUSTER_OVERVIEW_URL, CLUSTER_SETTINGS_URL, CLUSTER_URL } from '@qovery/shared/routes'
 import { Badge, ErrorBoundary, Header, Icon, Section, Skeleton, Tabs } from '@qovery/shared/ui'
 import NeedRedeployFlag from '../need-redeploy-flag/need-redeploy-flag'
 
 export function Container({ children }: PropsWithChildren) {
   const { organizationId = '', clusterId = '' } = useParams()
   const { pathname } = useLocation()
+  const isRunningStatusEnabled = useFeatureFlagVariantKey('cluster-running-status')
```
Comment (rmnbrd) [inline libs/pages/cluster/src/lib/ui/container/container.tsx]:
```suggestion
  const isClusterOverviewEnabled = useFeatureFlagVariantKey('cluster-running-status')
```

File: libs/shared/util-js/src/lib/resource-calculations.spec.ts [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    expect(formatNumber(1.2, 0)).toBe(1)
+  })
+
+  it('should use default precision of 2 when not specified', () => {
+    expect(formatNumber(1.2345)).toBe(1.23)
+    expect(formatNumber(1.2)).toBe(1.2)
+  })
+
+  it('should handle NaN and return 0', () => {
+    expect(formatNumber(NaN)).toBe(0)
```
Comment (rmnbrd) [inline libs/shared/util-js/src/lib/resource-calculations.spec.ts:30]: Nice! 👏

File: libs/domains/cluster-metrics/feature/src/lib/hooks/use-cluster-metrics-socket/use-cluster-metrics-socket.ts [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    url: QOVERY_WS + '/cluster/metrics',
+    urlSearchParams: {
+      organization: organizationId,
+      cluster: clusterId,
+    },
+    onMessage(queryClient, message: ClusterMetricsDto) {
+      // XXX: Filter out Fargate nodes (nodes without instance_type)
+      const filteredMessage = {
+        ...message,
+        nodes: message.nodes?.filter((node) => node.instance_type),
```
Comment (rmnbrd) [inline libs/domains/cluster-metrics/feature/src/lib/hooks/use-cluster-metrics-socket/use-cluster-metrics-socket.ts:21]: @RemiBonnet Out of curiosity, could this be done on the back-end side? Or is there other scenarios where we will need to consume all the events, including the ones that do not have an `instance_type`?

File: libs/domains/cluster-metrics/feature/src/lib/cluster-table-node/cluster-table-node.tsx [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+        <div className="flex h-8 w-1/4 items-center px-3">Nodes</div>
+        <div className="flex h-8 w-1/4 items-center px-3">CPU</div>
+        <div className="flex h-8 w-1/4 items-center px-3">Memory</div>
+        <div className="flex h-8 w-[calc(35%/3)] items-center px-3">Instance</div>
+        <div className="flex h-8 w-[calc(20%/3)] items-center px-3">Disk</div>
+        <div className="flex h-8 w-[calc(20%/3)] items-center px-3">Age</div>
+      </div>
+
+      {nodes
+        ?.sort((a, b) => new Date(b.created_at).getTime() - new Date(a.created_at).getTime())
```
Comment (rmnbrd) [inline libs/domains/cluster-metrics/feature/src/lib/cluster-table-node/cluster-table-node.tsx]: @RemiBonnet Could you merge this sorting logic with the ternary condition at [line 139](https://github.com/Qovery/console/pull/1914/files#diff-f7f42d4bd597232655ebc8b24e1690ec55ce6b7a8f17963ee9e0acac36660d6fR139) and extract it to a `useMemo` ? 🙏

File: libs/domains/cluster-metrics/feature/src/lib/cluster-card-node-usage/cluster-card-node-usage.tsx [hunk 7]
```diff
@@ -26,17 +30,17 @@ export function ClusterCardNodeUsage({ organizationId, clusterId }: ClusterCardN
...
       (node) => node.conditions?.find((condition) => condition.type === 'Ready')?.status === 'True'
     ).length || 0
 
   const warningNodes = Object.keys(runningStatus?.computed_status?.node_warnings || {}).length || 0
 
   const deployingNodes =
-    runningStatus?.nodes?.filter(
+    metrics?.nodes?.filter(
       (node) => node.conditions?.find((condition) => condition.type === 'Ready')?.status === 'False'
     ).length || 0
```
Comment (rmnbrd) [inline libs/domains/cluster-metrics/feature/src/lib/cluster-card-node-usage/cluster-card-node-usage.tsx]: @RemiBonnet Seeing as the value of `metrics` will often change, could you wrap these variables with a `useMemo` ? 🙏

File: libs/domains/cluster-metrics/feature/src/lib/hooks/use-cluster-metrics-socket/use-cluster-metrics-socket.ts (hunk 5 above)
Comment (RemiBonnet) [reply libs/domains/cluster-metrics/feature/src/lib/hooks/use-cluster-metrics-socket/use-cluster-metrics-socket.ts:21]: If I remember correctly, they want to keep it in the request for now, as it could be useful for debugging on their side! But I think it's something we could improve yes 👍

File: libs/domains/cluster-metrics/feature/src/lib/cluster-card-node-usage/cluster-card-node-usage.tsx (hunk 7 above)
Comment (RemiBonnet) [reply libs/domains/cluster-metrics/feature/src/lib/cluster-card-node-usage/cluster-card-node-usage.tsx]: Done! https://github.com/Qovery/console/pull/1914/commits/8ae33335dedae5cb6e03af880eeb12b4950d419d

File: libs/domains/cluster-metrics/feature/src/lib/cluster-table-node/cluster-table-node.tsx (hunk 6 above)
Comment (RemiBonnet) [reply libs/domains/cluster-metrics/feature/src/lib/cluster-table-node/cluster-table-node.tsx]: Done! https://github.com/Qovery/console/pull/1914/commits/8ae33335dedae5cb6e03af880eeb12b4950d419d

File: libs/pages/cluster/src/lib/ui/container/container.tsx (hunk 3 above)
Comment (RemiBonnet) [reply libs/pages/cluster/src/lib/ui/container/container.tsx]: https://github.com/Qovery/console/pull/1914/commits/8ae33335dedae5cb6e03af880eeb12b4950d419d

File: libs/domains/clusters/feature/src/lib/cluster-card/cluster-card.tsx (hunk 2 above)
Comment (RemiBonnet) [reply libs/domains/clusters/feature/src/lib/cluster-card/cluster-card.tsx]: https://github.com/Qovery/console/pull/1914/commits/8ae33335dedae5cb6e03af880eeb12b4950d419d

File: libs/domains/cluster-metrics/feature/src/lib/cluster-table-nodepool/calculate-nodepool-metrics.spec.ts (hunk 1 above)
Comment (RemiBonnet) [reply libs/domains/cluster-metrics/feature/src/lib/cluster-table-nodepool/calculate-nodepool-metrics.spec.ts:180]: Done! https://github.com/Qovery/console/pull/1914/commits/8ae33335dedae5cb6e03af880eeb12b4950d419d

**Review by rmnbrd — APPROVED** [review]
LGTM! Well done! 💯
